File: C++/EJEMPLOS/Cap1-CB/EJ1-9.c
/* Ejemplo Ej1-9 de Curso Basico de Fortran */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
int iecu;
#define MIN(a,b) ((a)<(b) ? (a) : (b))
#define MAX(a,b) ((a)>(b) ? (a) : (b))

double fun (double);
void bisec (double (*f) (double), double, double, double, double,
      int*, double*, double*, int*);

void main (void) {
extern int iecu;

int indice, niter;
double a, b, tolx, tolf, x, fx;

/* Descripcion de la convergencia alcanzada */

char estado[50];
struct casoconv {char *estado;} caso[]={
      "fa y fb tienen el mismo signo",
      "se ha alcanzado la precision deseada",
      "solo se ha alcanzado la precision en x",
      "solo se ha alcanzado la precision en f",
      "no se ha alcanzado la precision en x ni en f"};

/* Lectura del intervalo, tolerancias e iteraciones */

for (iecu=1; iecu<=2; iecu++) {

  printf ("\n%s%d", "Ecuacion ", iecu);
  printf ("\n%s", " Extremo a = "); scanf ("%lf",&a);
  printf ("%s", " Extremo b = ");  scanf ("%lf",&b);
  printf ("%s", " Tolerancia para x = "); scanf ("%lf",&tolx);
  printf ("%s", " Tolerancia para f = "); scanf ("%lf",&tolf);
  printf ("%s", " Iteraciones = "); scanf ("%d",&niter);

/* Ecuacion 1: En [0,1] hay solucion */
/* Ecuacion 2: En [4,8] hay solucion */

  bisec (fun, a, b, tolx, tolf, &niter, &x, &fx, &indice);

  if (indice == 0)
    printf ("\n%s\n", caso[0].estado);
  else
    printf ("\n%s%25.15lf\n%s%25.15lf\n%s\n%s%d\n", "x = ", x, "f = ",
            fx, caso[indice].estado, "Evaluaciones de f = ", niter);
  }
}

/* ---------------------------------------------------------------------
! PROPOSITO: Resolver la ecuacion f(x)=0 por el metodo de la biseccion

! FUNCIONES LLAMADAS:
!    f <=> ecuacion a resolver f(x)=0

!    La descripcion de f debe codificarse como sigue:

            double f (double x);

! ARGUMENTOS DE ENTRADA:
!    double a, b <=> extremos del intervalo
!    double tolx, tolf <=> tolerancias en x, f

! ARGUMENTOS DE SALIDA:
!    double x <=> valor estimado de la raiz
!    double fx <=> f(x)
!    int indice <=> situacion alcanzada
!        indice=0 : f(a) y f(b) tienen el mismo signo
!        indice=1 : se han alcanzado ambas tolerancias
!        indice=2 : solo se ha alcanzado la tolerancia en x
!        indice=3 : solo se ha alcanzado la tolerancia en f
!        indice=4 : no se ha alcanzado la tolerancia en x ni en f

! ARGUMENTOS DE ENTRADA/SALIDA:
!    int niter
!        En entrada: maximo numero de divisiones del intervalo
!        En salida : numero de evaluaciones de f realizadas
! ------------------------------------------------------------------- */

void bisec (double (*f) (double), double a, double b, double tolx,
       double tolf, int *niter, double *x, double *fx, int *indice) {

/* Variables locales */

int n;
double xx, x1, x2, f1, f2;

/* ----------------------------
 Primera sentencia ejecutable */

x1 = MIN(a,b) ; f1 = f(x1);
x2 = MAX(a,b) ; f2 = f(x2);
*niter = MAX(1,MIN(abs(*niter),50));

*indice = 0;
if (f1*f2 > 0) {   /* f(a) y f(b) tienen el mismo signo */
  *niter = 2; return; }

/* Se realizaran a lo sumo niter iteraciones */

for (n=1; n<=*niter; n++) {
  xx = x1 + (x2-x1)/2; *fx = f(xx);
  if (f1*(*fx) > 0)
    { x1 = xx ; f1 = *fx; }
  else
    { x2 = xx ; f2 = *fx; }
  if (x2-x1<tolx && fabs(*fx)<tolf)
    { *niter = 2 + n;
      *indice = 1;  break; }  /* se han alcanzado ambas tolerancias */
}

if (*indice == 0) { /* No se han alcanzado las dos tolerancias */
  *niter = 2 + *niter;
  if (x2-x1 < tolx)
    *indice = 2;   /* solo se ha alcanzado la tolerancia en x */
  else if (MIN(fabs(f1),fabs(f2)) < tolf)
    *indice = 3;   /* solo se ha alcanzado la tolerancia en f */
  else
    *indice = 4;   /* no se ha alcanzado la tolerancia en x ni en f */
}

/* Mejor punto obtenido */

if (fabs(f1) <= fabs(f2))
  { *x = x1 ; *fx = f1; }
else
  { *x = x2 ; *fx = f2; }
return;
}
/* ------------------------------------------------------------------ */
/* Ecuacion a resolver fun(x)=0  */

double fun (double x) {
double ff;
extern int iecu;

switch (iecu) {
case 1:
  ff = 1 + x*(-6+x*(15+x*(-24+x*(15+x*(-6+x))))) ;
  break;
case 2:
  ff = pow(x,4) + 2*pow(x,3) - 36*pow(x,2) - 2*x - 1;
}
return ff;
}

File: EJ1/EJ1-9.f90
MODULE numeroecu
  INTEGER iecu
ENDMODULE numeroecu

! ======================================================================
! PROPOSITO: Resolver la ecuacion f(x)=0 por el metodo de la biseccion

! SUBPROGRAMAS LLAMADOS:
!    DOUBLE PRECISION, EXTERNAL:: f <=> ecuacion a resolver f(x)=0

!    La funcion f(x) debe codificarse como sigue:

!           FUNCTION f(x)
!           DOUBLE PRECISION f, x
!             ...
!           f =
!           END

!    y declararse EXTERNAL en la unidad de programa de llamada.

! ARGUMENTOS DE ENTRADA:
!    DOUBLE PRECISION:: a, b <=> extremos del intervalo
!    DOUBLE PRECISION:: tolx, tolf <=> tolerancias en x, f

! ARGUMENTOS DE SALIDA:
!    DOUBLE PRECISION:: x <=> valor estimado de la raiz
!    DOUBLE PRECISION:: fx <=> f(x)
!    INTEGER:: indice <=> situacion alcanzada
!              indice=0 : f(a) y f(b) tienen el mismo signo
!              indice=1 : se han alcanzado ambas tolerancias
!              indice=2 : solo se ha alcanzado la tolerancia en x
!              indice=3 : solo se ha alcanzado la tolerancia en f
!              indice=4 : no se ha alcanzado la tolerancia en x ni en f

! ARGUMENTOS DE ENTRADA/SALIDA:
!    INTEGER:: niter
!              En entrada: maximo numero de divisiones del intervalo
!              En salida : numero de evaluaciones de f realizadas
! ======================================================================

SUBROUTINE bisec (f, a, b, tolx, tolf, niter, x, fx, indice)
IMPLICIT NONE

! Subprogramas llamados

DOUBLE PRECISION, EXTERNAL:: f

! Argumentos

DOUBLE PRECISION, INTENT(IN):: a, b, tolx, tolf
DOUBLE PRECISION, INTENT(OUT):: x, fx
INTEGER, INTENT(OUT):: indice
INTEGER, INTENT(INOUT):: niter

! Variables locales

INTEGER n
DOUBLE PRECISION x1, x2, f1, f2

! ----------------------------
! Primera sentencia ejecutable

x1 = MIN(a,b) ; f1 = f(x1)
x2 = MAX(a,b) ; f2 = f(x2)
niter = MAX(1,MIN(ABS(niter),50))

indice = 0
IF (f1*f2 > 0) THEN     ! f(a) y f(b) tienen el mismo signo
  niter = 2
  RETURN
ENDIF

! Se realizaran a lo sumo niter iteraciones

DO n = 1, niter
  x = 0.5d0*(x1+x2) ; fx = f(x)
  IF (f1*fx > 0) THEN
    x1 = x ; f1 = fx
  ELSE
    x2 = x ; f2 = fx
  ENDIF
  IF (x2-x1<tolx .AND. ABS(fx)<tolf) THEN
    niter = 2 + n
    indice = 1   ! se han alcanzado ambas tolerancias
    EXIT
  ENDIF
ENDDO

IF (indice == 0) THEN  ! No se han alcanzado las dos tolerancias
  niter = 2 + niter
  IF (x2-x1 < tolx) THEN
    indice = 2   ! solo se ha alcanzado la tolerancia en x
  ELSEIF (MIN(ABS(f1),ABS(f2)) < tolf) THEN
    indice = 3   ! solo se ha alcanzado la tolerancia en f
  ELSE
    indice = 4   ! no se ha alcanzado la tolerancia en x ni en f
  ENDIF
ENDIF

! Mejor punto obtenido

IF (ABS(f1) <= ABS(f2)) THEN
  x = x1 ; fx = f1
ELSE
  x = x2 ; fx = f2
ENDIF

ENDSUBROUTINE bisec

! ======================================================================
! Ecuacion a resolver fun(x)=0

FUNCTION fun(x)
USE numeroecu
DOUBLE PRECISION fun, x

SELECT CASE (iecu)
CASE (1)
  fun = 1 + x*(-6+x*(15+x*(-24+x*(15+x*(-6+x)))))
CASE (2)
  fun = x**4 + 2*x**3 - 36*x**2 - 2*x - 1
ENDSELECT

ENDFUNCTION fun

! ======================================================================
! Programa principal

PROGRAM ecuacion
USE numeroecu
IMPLICIT NONE

INTEGER indice, niter
CHARACTER(LEN=50) caso(0:4)
DOUBLE PRECISION a, b, tolx, tolf, x, fx
DOUBLE PRECISION, EXTERNAL:: fun

! Descripcion de la convergencia alcanzada

caso(0) = 'f(a) y f(b) tienen el mismo signo'
caso(1) = 'se ha alcanzado la precision deseada'
caso(2) = 'solo se ha alcanzado la precision en x'
caso(3) = 'solo se ha alcanzado la precision en f'
caso(4) = 'no se ha alcanzado la precision en x ni en f'

! Lectura del intervalo, tolerancias e iteraciones

DO iecu = 1, 2

  WRITE (*,'(/A,I2)') 'Ecuacion ', iecu
  WRITE (*,'(A)',ADVANCE='NO') ' Extremo a = ' ; READ*, a
  WRITE (*,'(A)',ADVANCE='NO') ' Extremo b = ' ; READ*, b
  WRITE (*,'(A)',ADVANCE='NO') ' Tolerancia para x = ' ; READ*, tolx
  WRITE (*,'(A)',ADVANCE='NO') ' Tolerancia para f = ' ; READ*, tolf
  WRITE (*,'(A)',ADVANCE='NO') ' Iteraciones = ' ; READ*, niter

  CALL bisec (fun, a, b, tolx, tolf, niter, x, fx, indice)

  IF (indice == 0) THEN
    WRITE (*,'(3X,A)') caso(0)
  ELSE
    WRITE (*,'(3X,A,G25.15/3X,A,G25.15/3X,A/3X,A,I2)') 'x = ', x,        &
             'f = ', fx, caso(indice), 'Evaluaciones de f = ', niter
  ENDIF

ENDDO

ENDPROGRAM ecuacion